documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,839 @@
+
+    CTL0 --> |Flow Miss| PreACL0
+    CTL0 --> |Flow Hit| AA0
+    PreACL0 --> MATS0
+    MATS0 --> AA0
+    AA0 --> |Flow Miss| PostACL0
+    PostACL0 --> CTU0
+    AA0 --> |Flow Hit| MS0
+    AA0 --> |Flow Change| CTU0
+    CTU0 --> MS0
+    MS0 --> Out
+
+    CTL1 --> |Flow Miss| PreACL1
+    CTL1 --> |Flow Hit| AA1
+    PreACL1 --> MATS1
+    MATS1 --> AA1
+    AA1 --> |Flow Miss| PostACL1
+    PostACL1 --> CTU1
+    AA1 --> |Flow Hit| MS1
+    AA1 --> |Flow Change| CTU1
+    CTU1 --> MS1
+    MS1 --> Out
+```
+
+This design allows our upper layers to be flexible and doesn't limit to any specific object models for modeling their own policies. For example, [SONiC-DASH pipeline](https://github.com/sonic-net/SONiC/blob/master/doc/dash/dash-sonic-hld.md#2-packet-flows) uses DASH-SAI pipeline to implement VNET routing scenarios by translating the [SONiC DASH APIs](https://github.com/sonic-net/sonic-dash-api/tree/master/proto) to DASH-SAI model:
+
+- A DASH-SAI pipeline is used to represent a VM NIC (ENI).
+- The VxLAN VNI is used to do the direction lookup.
+- The inner MAC address is used for pipeline lookup, a.k.a. ENI lookup (ENI). 
+- Once it goes into the corresponding DASH pipeline, the `outbound` pipeline will be used to process the packets coming from the VM, while the `inbound` pipeline will be used to process the packets going into the VM.
+
+## 5. Pipeline components
+
+### 5.1. Per-packet metadata bus
+
+First of all, since we have multiple matching stages in the pipeline, we need a way to pass the information from the matched entries in the earlier stages to the later ones to help us making final decisions on packet transformation. And this is what metadata bus is for.
+
+At high-level, the metadata bus is a set of fields that being carried all the way through the pipeline along with the packet. It contains:
+
+- The information from the original packet, such as encap information.
+- The information from each matched entry and related data structures, e.g., when a VNET mapping entry is matched, we will publish the information from VNET to the metadata bus. 
+- The routing action and its inline parameters from the matched entry.
+
+Implementation-wise, this is similar to Packet Header Vector or Bus in NPL.
+
+### 5.2. Direction Lookup
+
+In DASH-SAI pipeline, traffic are split into 2 directions: `inbound` and `outbound`. Each direction has its own pipeline (see pipeline overview above). When a new packet arrives, we will assign a direction to the packet, then process the packet in the corresponding pipeline. This ensures us to match the flow and transform the packet in the right way.
+
+### 5.3. Pipeline Lookup
+
+DASH supports multi-tenancy model for traffic handling. A single device can have multiple pipelines, and each pipeline is used to handle traffic for a specific tenant. When a packet arrives, besides direction lookup, we also need pipeline lookup to determine which pipeline to use for processing the packet.
```

Comment:
Nit: You may want to mention here that "each is pipeline is used to handle traffic for a specific tenant OR instance of a service (L4LB)"

Reply:
right! this is actually a good one. Since LB is also just an example, I have updated this part to " or whichever network function being modeled, such as a load balancer".

Reply:
updated now.